File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/ime/KanaModifier.kt
package com.draftright.keyboard.ime

/**
 * The 小゛゜ modifier for Japanese flick (#212, phase 3): cycles the last kana
 * through its small / dakuten / handakuten variants, the way the 大⇔小 key on a
 * real flick keyboard does — か→が→か, は→ば→ぱ→は, つ→っ→づ→つ, や→ゃ→や.
 * Without it じ/ば/ぱ/small kana can't be typed at all.
 *
 * The variant cycles are a cited linguistic constant (one source of truth,
 * mirror of the Swift `KanaModifier`, parity-guarded). Kana with no variants
 * (な/ま/ら rows, ん, を) return unchanged.
 */
object KanaModifier {

    /** Each row = a kana and its variants, in cycle order (base first). */
    private val cycles: List<List<String>> = listOf(
        listOf("あ", "ぁ"), listOf("い", "ぃ"), listOf("う", "ぅ", "ゔ"), listOf("え", "ぇ"), listOf("お", "ぉ"),
        listOf("か", "が"), listOf("き", "ぎ"), listOf("く", "ぐ"), listOf("け", "げ"), listOf("こ", "ご"),
        listOf("さ", "ざ"), listOf("し", "じ"), listOf("す", "ず"), listOf("せ", "ぜ"), listOf("そ", "ぞ"),
        listOf("た", "だ"), listOf("ち", "ぢ"), listOf("つ", "っ", "づ"), listOf("て", "で"), listOf("と", "ど"),
        listOf("は", "ば", "ぱ"), listOf("ひ", "び", "ぴ"), listOf("ふ", "ぶ", "ぷ"), listOf("へ", "べ", "ぺ"), listOf("ほ", "ぼ", "ぽ"),
        listOf("や", "ゃ"), listOf("ゆ", "ゅ"), listOf("よ", "ょ"),
        listOf("わ", "ゎ"),
    )

    private val next: Map<String, String> = buildMap {
        for (cycle in cycles) for (i in cycle.indices) put(cycle[i], cycle[(i + 1) % cycle.size])
    }

    /** The next variant of [kana] in its cycle, or [kana] unchanged if it has none. */
    fun cycle(kana: String): String = next[kana] ?: kana
}

File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/IME/KanaModifier.swift
import Foundation

/// The 小゛゜ modifier for Japanese flick (#212, phase 3): cycles the last kana
/// through its small/dakuten/handakuten variants — か→が→か, は→ば→ぱ→は,
/// つ→っ→づ→つ, や→ゃ→や. Cited linguistic constant, mirror of the Kotlin
/// `KanaModifier` (parity-guarded). Kana with no variants return unchanged.
public enum KanaModifier {

    private static let cycles: [[String]] = [
        ["あ", "ぁ"], ["い", "ぃ"], ["う", "ぅ", "ゔ"], ["え", "ぇ"], ["お", "ぉ"],
        ["か", "が"], ["き", "ぎ"], ["く", "ぐ"], ["け", "げ"], ["こ", "ご"],
        ["さ", "ざ"], ["し", "じ"], ["す", "ず"], ["せ", "ぜ"], ["そ", "ぞ"],
        ["た", "だ"], ["ち", "ぢ"], ["つ", "っ", "づ"], ["て", "で"], ["と", "ど"],
        ["は", "ば", "ぱ"], ["ひ", "び", "ぴ"], ["ふ", "ぶ", "ぷ"], ["へ", "べ", "ぺ"], ["ほ", "ぼ", "ぽ"],
        ["や", "ゃ"], ["ゆ", "ゅ"], ["よ", "ょ"],
        ["わ", "ゎ"],
    ]

    private static let next: [String: String] = {
        var m: [String: String] = [:]
        for cycle in cycles {
            for i in cycle.indices { m[cycle[i]] = cycle[(i + 1) % cycle.count] }
        }
        return m
    }()

    /// The next variant of `kana`, or `kana` unchanged if it has none.
    public static func cycle(_ kana: String) -> String { next[kana] ?? kana }
}
